accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,423 @@
+    "package-b": "1.0.0"
+  },
+  "resolutions": {
+    "package-a/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` oly for `package-a` and `package-b` will
+still have `package-d1@2.0.0` in its own `node_modules`.
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-a": "3.0.0"
+  }
+```
+
+`package-a` will still be resolved to `1.0.0`, but `package-c` will have
+`package-a@3.0.0` in its own `node_modules`.
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "package-a": "3.0.0"
+  }
+```
+
+yarn will do nothing (see below why).
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-a/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` both for `package-a` and the nested
+dependency `package-a` of `package-c`.
+
+## Resolutions
+
+Each sub-field of the `resolutions` field is called a *resolution*.
+It is a JSON field expressed by two strings: the package designation on the
+left and a version specification on the right.
+
+### Package designation
+
+A *resolution* contains on the left-hand side a glob pattern applied to
+the dependency tree (and not to the `node_modules` directory tree, since the
+latter is the result of yarn resolution being influenced by the *resolution*).
+
+- `a/b` denotes the directly nested dependency `b` of the project's
+dependency `a`.
+- `**/a/b` denotes the directly nested dependency `b`
+of all the dependencies and nested dependencies `a` of the project.
+- `a` denotes the project's dependency `a` (see below for a discussion on
+the matter of specifying a *resolution* for one of the project dependency).
+- `a/**/b` denotes all the nested dependencies `b` of the project's
+dependency `a`.
+- `**/a` denotes all the nested dependencies `a` of the project.
+- `**/a-*` denotes all the nested dependencies of the project whose
+name starts with `a-`.
+- `**` denotes all the nested dependencies of the project (a bad idea mostly,
+as well as all other designations ending with `**`).
+
+### Version specification
+
+A *resolution* contains on the right-hand side a version specification
+interpreted via the `semver` package as usually done in yarn.
+
+## Relation to non-nested dependencies
+
+The `devDependencies` and `dependencies` fields always take precedence over the
+`resolutions` field: if the user defines explicitly a dependency there,
+it means that he wants that version, even if it's specified with a non-exact
+specification. So the `resolutions` field only applies to nested-dependencies.
+Nevertheless, in case of incompatibility between the specification of a
+non-nested dependency version and a *resolution*, a warning is issued.
+
+## Relation to the `--flat` option
+
+The `--flat` option becomes thus a way to populate the resolutions field for

```

Comment:
With this RFC, `--flat` becomes a flag that have an impact only on modifying the `package.json`, at the same level as `add`, `upgrade` or `upgrade-interactive` do.

It is not used anymore for `install` at all: `install` uses whatever is in the `resolutions` field (and the content of the later could have been populated manually or via the use of `--flat`).

But actually your question makes it clear that it could be useful to take `--flat` into accounts in checks too, to ensure that all dependencies of the project are actually flattened (via the `resolutions` field) as expected. I will add it to the RFC :)